### PR TITLE
routerrpc+itest: add validation to MPP params

### PR DESCRIPTION
**Pull request by NishantBansal2003**
## Change Description
Fixes: #8916 
Adds validation to ensure that MPP parameters are compatible with the payment amount before attempting the payment. This prevents payments from entering a path finding loop that would eventually timeout.

## Steps to Test
Added itest to validate that a payment with multipath parameters exceeding the allowed maximum amount fails with the expected error.

### Testing
- [x] Your PR passes all CI checks.
- [x] Tests covering the positive and negative (error paths) are included.
- [x] Bug fixes contain tests triggering the bug to prevent regressions.

### Code Style and Documentation
- [x] The change is not [insubstantial](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#substantial-contributions-only). Typo fixes are not accepted to fight bot spam.
- [x] The change obeys the [Code Documentation and Commenting](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#code-documentation-and-commenting) guidelines, and lines wrap at 80.
- [x] Commits follow the [Ideal Git Commit Structure](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#ideal-git-commit-structure).
- [ ] Any new logging statements use an appropriate subsystem and logging level.
- [ ] Any new lncli commands have appropriate tags in the comments for the rpc in the proto file.
- [x] [There is a change description in the release notes](https://github.com/lightningnetwork/lnd/tree/master/docs/release-notes), or `[skip ci]` in the commit message for small changes.

📝 Please see our [Contribution Guidelines](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md) for further guidance.

### CODE REVIEW

File: itest/lnd_mpp_test.go [hunk 1]
```diff
@@ -61,9 +96,10 @@ func testSendMultiPathPayment(ht *lntest.HarnessTest) {
 	payReq := payReqs[0]
 
 	sendReq := &routerrpc.SendPaymentRequest{
-		PaymentRequest: payReq,
-		MaxParts:       10,
-		FeeLimitMsat:   noFeeLimitMsat,
+		PaymentRequest:   payReq,
+		MaxParts:         10,
+		MaxShardSizeMsat: 30_000_000,
```
Comment (yyforyongyu) [inline itest/lnd_mpp_test.go:101]: why do we need to set `MaxShardSizeMsat` here?

File: itest/lnd_mpp_test.go [hunk 2]
```diff
@@ -13,10 +13,42 @@ import (
...
 	"github.com/lightningnetwork/lnd/lntypes"
+	"github.com/lightningnetwork/lnd/lnwire"
 	"github.com/lightningnetwork/lnd/routing/route"
 	"github.com/stretchr/testify/require"
 )
 
+// testValidateMPPParams validates that a payment with multipath parameters
+// exceeding the allowed maximum amount fails with the expected error.
+func testValidateMPPParams(
+	ht *lntest.HarnessTest, mts *mppTestScenario, paymentAmt btcutil.Amount,
```
Comment (yyforyongyu) [inline itest/lnd_mpp_test.go:24]: the format is not quite right, see https://github.com/lightningnetwork/lnd/blob/master/docs/code_formatting_rules.md#wrapping-long-function-definitions

File: itest/lnd_mpp_test.go [hunk 3]
```diff
@@ -53,6 +85,9 @@ func testSendMultiPathPayment(ht *lntest.HarnessTest) {
 		mts.alice, mts.dave, expectedPolicy, chanPointAliceDave, false,
 	)
 
+	// Validate that multipath payment parameters are enforced correctly.
+	testValidateMPPParams(ht, mts, paymentAmt)
```
Comment (yyforyongyu) [inline itest/lnd_mpp_test.go:89]: Also check this in `testSendToRouteMultiPath`?

**Review by yyforyongyu — COMMENTED** [review]
Thanks for the PR! Left a few comments re the format and test.

File: itest/lnd_mpp_test.go (hunk 1 above)
Comment (NishantBansal2003) [reply itest/lnd_mpp_test.go:101]: To validate that when the mpp params permit sending the full payment amount, the payment is successfully settled. I'll add a comment in the test to make this clear.

**Review by yyforyongyu — APPROVED** [review]
LGTM🌹

File: lnrpc/routerrpc/router_backend.go [hunk 4]
```diff
@@ -1191,6 +1191,21 @@ func (r *RouterBackend) extractIntentFromSendRequest(
 		payIntent.DestFeatures = features
 	}
 
+	// Validate that the MPP parameters are compatible with the
```
Comment (ziggie1984) [inline lnrpc/routerrpc/router_backend.go:1194]: Nice validation, I think we can even add more strict rules:

1. Add an upper limit for the `MaxParts` like maybe 100
2. Maybe we can also suggest the minumum number of shards as a user feedback so he knows what to set.
```
  minRequiredParts := (*payIntent.Amount + *payIntent.MaxShardAmt - 1) / *payIntent.MaxShardAmt

```

File: itest/lnd_mpp_test.go [hunk 5]
```diff
@@ -13,10 +13,41 @@ import (
...
+	"github.com/lightningnetwork/lnd/lnwire"
 	"github.com/lightningnetwork/lnd/routing/route"
 	"github.com/stretchr/testify/require"
 )
 
+// testValidateMPPParams validates that a payment with multipath parameters
+// exceeding the allowed maximum amount fails with the expected error.
+func testValidateMPPParams(ht *lntest.HarnessTest, mts *mppTestScenario,
+	paymentAmt btcutil.Amount) {
+	// Create an invoice from Bob for the payment.
```
Comment (ziggie1984) [inline itest/lnd_mpp_test.go:25]: Nit: New line above

File: itest/lnd_mpp_test.go (hunk 1 above)
Comment (ziggie1984) [reply itest/lnd_mpp_test.go:99]: Instead of just adding here the `MaxShardSizeMsat ` please create a separate test which tests exactly the validation. This separates the test cases nicely and can be build integrating your test above `testValidateMPPParams`

**Review by ziggie1984 — CHANGES_REQUESTED** [review]
Thanks for the fix, left some comments.

File: lnrpc/routerrpc/router_backend.go (hunk 4 above)
Comment (NishantBansal2003) [reply lnrpc/routerrpc/router_backend.go:1194]: Certainly, I have been trying to understand what upper limit can be set for `MaxParts`. However, I am uncertain why a value of `100` was proposed, as certain sections of the code—particularly in the tests—use the value `1000`. Could you please clarify this matter so that I may gain a better understanding?

Additionally, when suggesting `minRequiredParts`, I am considering an edge case where the value exceeds `100`. In that scenario, should I also return the error that indicates we need to increase the `MaxShardAmt`?

Comment (ziggie1984) [reply lnrpc/routerrpc/router_backend.go:1194]: My idea behind restricting the maximum shard part was, that we do not want a user to launch a payment with the wrong values and all of a sudden create like 1000 payments. Looking at the code we use the max limit of 1000 but I don't think we ever hit this limit, looks like we can easily come up with a lower value for the test case but we need to verify. 
So maybe have a higher value like 1000 shard parts ? I mean if a payment has to use more that 1000. shards I would be very concerned. 

I would say for the edge case that the minRequiredParts exceeds the max shard number of 1000 we might return an error that its not possible to send this amount with this MaxPartAmt and MaxParts, I think it's not necessary to tell the caller that he has to increase the MaxShardAmt value.

File: lnrpc/routerrpc/router_backend.go [hunk 6]
```diff
@@ -1191,6 +1195,46 @@ func (r *RouterBackend) extractIntentFromSendRequest(
...
+	// payment amount. In other words, the parameters are invalid if
+	// they do not permit sending the full payment amount.
+	if rpcPayReq.MaxShardSizeMsat > 0 {
+		if payIntent.MaxParts > MaxPartsUpperLimit {
```
Comment (yyforyongyu) [inline lnrpc/routerrpc/router_backend.go:1202]: this check should be moved to where we assign `payIntent.MaxParts` at L863 so we can exit early.

File: lnrpc/routerrpc/router_backend.go [hunk 7]
```diff
@@ -1191,6 +1195,46 @@ func (r *RouterBackend) extractIntentFromSendRequest(
 		payIntent.DestFeatures = features
 	}
 
+	// Validate that the MPP parameters are compatible with the
+	// payment amount. In other words, the parameters are invalid if
+	// they do not permit sending the full payment amount.
+	if rpcPayReq.MaxShardSizeMsat > 0 {
```
Comment (yyforyongyu) [inline lnrpc/routerrpc/router_backend.go:1201]: think we should check `payIntent.MaxShardAmt` for clarity.

File: lnrpc/routerrpc/router_backend.go (hunk 4 above)
Comment (yyforyongyu) [reply lnrpc/routerrpc/router_backend.go:1198]: I think `minRequiredParts` can be confusing as it suggests one way to fix the case, while it's a function of two params `MaxShardAmt*MaxParts`. The user can choose to increase either the amt per shard or the num of shards so I'd say leave it to the user to decide. Plus when suggesting `minRequiredParts` it implicitly drives the number of HTLC attempts to be used, which can cause more failure in routing.

File: itest/lnd_mpp_test.go [hunk 8]
```diff
@@ -13,10 +14,124 @@ import (
...
+				lnwire.MilliSatoshi(paymentAmt*1000)),
+		},
+	}
+
+	for _, tc := range testCases {
+		test := tc
+
+		success := ht.Run(test.name, func(t *testing.T) {
+			// Bob creates an invoice for the payment.
+			payReqs, _, _ := ht.CreatePayReqs(mts.bob, paymentAmt,
```
Comment (yyforyongyu) [inline itest/lnd_mpp_test.go:80]: I think we can instead create a single invoice for the whole test case as we know only the last send payment will succeed.

**Review by ziggie1984 — CHANGES_REQUESTED** [review]
I think we should just add a unit-test instead of adding a itest.

Moreover let's remove the minRequired parts check yy has a valid point.

File: lnrpc/routerrpc/router_backend.go [hunk 9]
```diff
@@ -1191,6 +1205,22 @@ func (r *RouterBackend) extractIntentFromSendRequest(
 		payIntent.DestFeatures = features
 	}
 
+	// Validate that the MPP parameters are compatible with the
+	// payment amount. In other words, the parameters are invalid if
```
Comment (ziggie1984) [inline lnrpc/routerrpc/router_backend.go:1209]: why aren't we doing this check also where we check the MaxShardParts ? We already have all the infos available earlier ?

File: lnrpc/routerrpc/router_backend_test.go [hunk 10]
```diff
@@ -475,3 +475,80 @@ func testUnmarshalAMP(t *testing.T, test unmarshalAMPTest) {
 		t.Fatalf("test case has non-standard outcome")
 	}
 }
+
+type ExtractIntentTestCase struct {
+	name             string
+	maxPartsInput    uint32
+	shardAmountMsat  uint64
```
Comment (ziggie1984) [inline lnrpc/routerrpc/router_backend_test.go:482]: Nit: Why not keep the naming here from the above code ?

`maxShardAmt, maxParts` ?

File: lnrpc/routerrpc/router_backend_test.go [hunk 11]
```diff
@@ -475,3 +475,80 @@ func testUnmarshalAMP(t *testing.T, test unmarshalAMPTest) {
...
+	valid            bool
+	expectedErrorMsg string
+}
+
```
Comment (ziggie1984) [inline lnrpc/routerrpc/router_backend_test.go:486]: Missing godoc, would be great if you could properly test the whole extract func. otherwise create a comment which specifies that you are in specific only testing the shard multipart payment logic.

File: lnrpc/routerrpc/router_backend_test.go [hunk 12]
```diff
@@ -475,3 +475,80 @@ func testUnmarshalAMP(t *testing.T, test unmarshalAMPTest) {
...
+			name: "Valid MPP parameters, payment " +
+				"settled",
+			maxPartsInput:   10,
+			shardAmountMsat: 30_000_000,
+			valid:           true,
+		},
+	}
+
+	for _, testCase := range testCases {
+		testCase := testCase
```
Comment (ziggie1984) [inline lnrpc/routerrpc/router_backend_test.go:523]: I think this iis not necessary anymore with the go compiler option loopvar

File: lnrpc/routerrpc/router_backend_test.go [hunk 13]
```diff
@@ -475,3 +475,80 @@ func testUnmarshalAMP(t *testing.T, test unmarshalAMPTest) {
...
+		t.Run(testCase.name, func(t *testing.T) {
+			runExtractIntentTest(t, backend, testCase)
```
Comment (ziggie1984) [inline lnrpc/routerrpc/router_backend_test.go:525]: add `t.Parallel() ` to speed up testing.

**Review by ziggie1984 — CHANGES_REQUESTED** [review]
Nice, left some comments.

File: lnrpc/routerrpc/router_backend.go (hunk 9 above)
Comment (ziggie1984) [reply lnrpc/routerrpc/router_backend.go:1209]: gottya

File: lnrpc/routerrpc/router_backend_test.go (hunk 11 above)
Comment (NishantBansal2003) [reply lnrpc/routerrpc/router_backend_test.go:487]: I have added tests for the entire extract function.

File: lnrpc/routerrpc/router_backend_test.go [hunk 14]
```diff
@@ -475,3 +476,348 @@ func testUnmarshalAMP(t *testing.T, test unmarshalAMPTest) {
...
+	backend          *RouterBackend
+	sendReq          *SendPaymentRequest
+	valid            bool
+	expectedErrorMsg string
+}
+
+// TestExtractIntentFromSendRequest verifies that extractIntentFromSendRequest
+// correctly translates a SendPaymentRequest from an RPC client into a
+// LightningPayment intent.
+func TestExtractIntentFromSendRequest(t *testing.T) {
```
Comment (yyforyongyu) [inline lnrpc/routerrpc/router_backend_test.go:491]: 👍

**Review by yyforyongyu — APPROVED** [review]
Nice test!

File: lnrpc/routerrpc/router_backend_test.go [hunk 15]
```diff
@@ -475,3 +476,348 @@ func testUnmarshalAMP(t *testing.T, test unmarshalAMPTest) {
 		t.Fatalf("test case has non-standard outcome")
 	}
 }
+
+type ExtractIntentTestCase struct {
```
Comment (MPins) [inline lnrpc/routerrpc/router_backend_test.go:480]: nit: this struct doesn't need to be exported, so it can start with a lowercase letter.

File: lnrpc/routerrpc/router_backend_test.go [hunk 16]
```diff
@@ -475,3 +476,348 @@ func testUnmarshalAMP(t *testing.T, test unmarshalAMPTest) {
...
+				LastHopPubkey:  []byte{1},
+			},
+			valid:            false,
+			expectedErrorMsg: "invalid vertex length",
+		},
+		{
+			name: "Max parts exceed allowed limit",
+			backend: &RouterBackend{
+				MaxTotalTimelock: 1000,
+			},
```
Comment (MPins) [inline lnrpc/routerrpc/router_backend_test.go:541]: nit: Setting MaxTotalTimelock is not needed.

File: lnrpc/routerrpc/router_backend_test.go [hunk 17]
```diff
@@ -475,3 +476,348 @@ func testUnmarshalAMP(t *testing.T, test unmarshalAMPTest) {
...
+			valid: false,
+			expectedErrorMsg: "requested max_parts (1001) exceeds" +
+				" the allowed upper limit",
+		},
+		{
+			name: "Fee limit conflict, both sat and msat specified",
+			backend: &RouterBackend{
+				MaxTotalTimelock: 1000,
+			},
```
Comment (MPins) [inline lnrpc/routerrpc/router_backend_test.go:554]: nit: Setting `MaxTotalTimelock` is not needed.

File: lnrpc/routerrpc/router_backend_test.go [hunk 18]
```diff
@@ -475,3 +476,348 @@ func testUnmarshalAMP(t *testing.T, test unmarshalAMPTest) {
...
+			valid: false,
+			expectedErrorMsg: "sat and msat arguments are " +
+				"mutually exclusive",
+		},
+		{
+			name: "Dest custom records with type below minimum",
+			backend: &RouterBackend{
+				MaxTotalTimelock: 1000,
+			},
```
Comment (MPins) [inline lnrpc/routerrpc/router_backend_test.go:567]: nit: Setting `MaxTotalTimelock` is not needed.

File: lnrpc/routerrpc/router_backend_test.go [hunk 19]
```diff
@@ -475,3 +476,348 @@ func testUnmarshalAMP(t *testing.T, test unmarshalAMPTest) {
...
+				},
+			},
+			valid:            false,
+			expectedErrorMsg: "no custom records with types below",
+		},
+		{
+			name: "Custom record entry with TLV type below minimum",
+			backend: &RouterBackend{
+				MaxTotalTimelock: 1000,
+			},
```
Comment (MPins) [inline lnrpc/routerrpc/router_backend_test.go:580]: nit: Setting `MaxTotalTimelock` is not needed.

**Review by MPins — COMMENTED** [review]
Well done :+1: 

Left some comments that might help make the test a bit cleaner.

File: lnrpc/routerrpc/router_backend_test.go [hunk 20]
```diff
@@ -475,3 +476,331 @@ func testUnmarshalAMP(t *testing.T, test unmarshalAMPTest) {
...
+		{
+			name:    "Invalid last hop pubkey",
+			backend: &RouterBackend{},
+			sendReq: &SendPaymentRequest{
+				OutgoingChanId: 38484,
+				LastHopPubkey:  []byte{1},
+			},
+			valid:            false,
+			expectedErrorMsg: "invalid vertex length",
+		},
```
Comment (MPins) [inline lnrpc/routerrpc/router_backend_test.go:536]: You can add the test below to increase the code coverage:
```
		{
			name: "total time lock exceeds max allowed",
			backend: &RouterBackend{
				MaxTotalTimelock: 1000,
			},
			sendReq: &SendPaymentRequest{
				CltvLimit: 1001,
			},
			valid: false,
			expectedErrorMsg: "total time lock of 1001 exceeds "+
			"max allowed 1000",
		},
```

File: lnrpc/routerrpc/router_backend_test.go [hunk 21]
```diff
@@ -475,3 +476,331 @@ func testUnmarshalAMP(t *testing.T, test unmarshalAMPTest) {
...
+				"specified",
+			backend: &RouterBackend{},
+			sendReq: &SendPaymentRequest{
+				FeeLimitSat:  1000000,
+				FeeLimitMsat: 1000000000,
+			},
+			valid: false,
+			expectedErrorMsg: "sat and msat arguments are " +
+				"mutually exclusive",
+		},
```
Comment (MPins) [inline lnrpc/routerrpc/router_backend_test.go:559]: You can add the test below to increase the code coverage:
```
		{
			name: "Fee limit cannot be negative",
			backend: &RouterBackend{},
			sendReq: &SendPaymentRequest{
				FeeLimitSat:  -1,
			},
			valid: false,
			expectedErrorMsg: "amount cannot be negative",
		},
```

**Review by MPins — COMMENTED** [review]
Looking closer, I found some missing tests and suggest a few naming changes to make it clearer what each test is actually doing.

Besides the suggested additions I made, we could also add a test for the case where the invoice has an amount specified and the payer is not allowed to override it. However, to test that properly, we would need a non-expired invoice.

**Comment by NishantBansal2003** [comment]
> we could also add a test for the case where the invoice has an amount specified and the payer is not allowed to override it. However, to test that properly, we would need a non-expired invoice.

Yes, we require a non-expired invoice, which is why I was unable to add a test for it. Perhaps, in such cases, "itest" would suffice?

**Review by MPins — APPROVED** [review]
LGTM

**Comment by MPins** [comment]
> suffice
Maybe creating an iTest is too much just to cover this use case.

**Comment by ziggie1984** [comment]
you can use: ` make unit-cover case=TestExtractIntentFromSendRequest` and check the result coverage file for `extractIntentFromSendRequest ` function to check if everything is covered using: ` go tool cover -html=coverage.txt`, thats a pretty nice check on your tests.

**Comment by NishantBansal2003** [comment]
> you can use: ` make unit-cover case=TestExtractIntentFromSendRequest` and check the result coverage file for `extractIntentFromSendRequest ` function to check if everything is covered using: ` go tool cover -html=coverage.txt`, thats a pretty nice check on your tests.

Thanks, this helped! I have also updated the tests for the latest changes.

**Review by ziggie1984 — APPROVED** [review]
LGTM 🎉

**Comment by guggero** [comment]
Hmm, CI seems to have gotten stuck on this one. Can you please push again to re-trigger (rebase on master for example)?

**Comment by guggero** [comment]
Rebased and pushed the commits to re-trigger CI, so we can merge this.
